cmd/containerd-shim-runhcs-v1/diag.go

```diff
@@ -0,0 +1,74 @@
+package main
+
+import (
+	"context"
+	"io"
+
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+	"github.com/Microsoft/hcsshim/internal/shimdiag"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+)
+
+type taskDiagnostics interface {
+	ExecInHost(ctx context.Context, req *shimdiag.ExecProcessRequest) (int, error)
+}
+
+func execInUvm(ctx context.Context, vm *uvm.UtilityVM, req *shimdiag.ExecProcessRequest) (int, error) {
+	np, err := newNpipeIO(ctx, "", "", req.Stdin, req.Stdout, req.Stderr, req.Terminal)
+	if err != nil {
+		return 0, err
+	}
+	wd := req.Workdir
+	if wd == "" {
+		if vm.OS() == "windows" {
+			wd = `c:\`
+		} else {
+			wd = "/"
+		}
+	}
+	proc, err := vm.ComputeSystem().CreateProcess(&hcsschema.ProcessParameters{
+		CommandArgs:      req.Args,
+		CreateStdInPipe:  req.Stdin != "",
+		CreateStdOutPipe: req.Stdout != "",
+		CreateStdErrPipe: req.Stderr != "",
+		EmulateConsole:   req.Terminal,
+		WorkingDirectory: wd,
+	})
+	if err != nil {
+		np.Close()
+		return 0, err
+	}
+	pin, pout, perr, err := proc.Stdio()
+	if err != nil {
+		np.Close()
+		return 0, err
+	}
+	if pin != nil {
+		go func() {
+			io.Copy(pin, np.Stdin())
+			pin.Close()
+			proc.CloseStdin()
+		}()
+	}
+	if pout != nil {
+		go func() {
+			io.Copy(np.Stdout(), pout)
+			pout.Close()
+		}()
+	}
+	if perr != nil {
+		go func() {
+			io.Copy(np.Stderr(), perr)
+			perr.Close()
+		}()
+	}
+	err = proc.Wait()
+	if err != nil {
+		proc.Kill()
+		proc.Wait()
+	}
+	ec, err := proc.ExitCode()
+	proc.Close()
+	np.Close()
```

Comment:
It looks like we `np.Close()` on all exit points from this function. Should we just `defer` it?

Reply:
Ah good call. In my original design, we did not wait for process completion...